docs/input_output.md

```diff
@@ -0,0 +1,196 @@
+the `dsub` command-line:
+
+```
+--output OUTPUT_FILE=gs://bucket/path/file.bam
+```
+
+Then have your job write the output file to the path
+`/mnt/data/output/gs/bucket/path/file.bam` within the Docker container.
+
+The file will be automatically copied to Cloud Storage when your script or
+command exits with success.
+
+The Docker container will receive the environment variable:
+
+```
+OUTPUT_FILE=/mnt/data/output/gs/bucket/path/file.bam
+```
+
+### 2. Copy a file pattern to Cloud Storage.
+
+To copy a set of files to Cloud Storage, specify the full URL pattern on
+the `dsub` command-line:
+
+```
+--output OUTPUT_FILES=gs://bucket/path/*.bam
+```
+
+Then have your job write output files to the path
+`/mnt/data/output/gs/bucket/path` within the Docker container.
+
+All files matching the pattern `/mnt/data/output/gs/bucket/path/*.bam` will be
+automatically copied to the Cloud Storage when your script or
+command exits with success.
+
+The Docker container will receive the environment variable:
+
+```
+OUTPUT_FILES=/mnt/data/output/gs/bucket/path/*.bam
+```
+
+Typically a job script will have the output file extensions hard-coded, but
+if needed it can be parsed from the environment variable. More commonly,
+the job script will need the output directory.
+
+To get the output directory and file extension in bash:
+
+```
+# This will set OUTPUT_DIR to "/mnt/data/output/gs/bucket/path"
+OUTPUT_DIR="$(dirname "${OUTPUT_FILES}")"
+
+# This will set OUTPUT_FILE_PATTERN to "*.bam"
+OUTPUT_FILE_PATTERN="$(basename "${OUTPUT_FILES}")"
+
+# This will set OUTPUT_EXTENSION to "bam" using the bash prefix removal
+# operator "##", matching the longest pattern up to and including the period.
+OUTPUT_EXTENSION="${OUTPUT_FILE_PATTERN##*.}"
+```
+
+### 3. Copy a directory, recursively to Cloud Storage.
+
+To recursively copy a directory of output to Cloud Storage, use the
+`dsub` command-line flag `--output-recursive`:
+
+```
+--output-recursive OUTPUT_PATH=gs://bucket/path
+```
+
+Then have your job write output files and subdirectories to the path
+`/mnt/data/output/gs/bucket/path` within the Docker container.
+
+The object(s) at the Cloud Storage path will be recursively copied and
+made available at the path `/mnt/data/input/gs/bucket/path`.
+
+The Docker container will receive the environment variable:
+
+```
+OUTPUT_PATH=/mnt/data/output/gs/bucket/path/
+```
+
+## Notice
+
+Note about the `--input-recursive` and `--output-recursive` flags
+
+As a getting started convenience, if `--input-recursive` or `--output-recursive`
+are used, dsub will automatically check for and, if needed, install the
+[Google Cloud SDK](https://cloud.google.com/sdk/docs/) in the Docker container
+at runtime (before your script executes).
+
+If you use the recursive copy features, you are encouraged to install gcloud
+in your Docker image when you build it in order to avoid the installation at
+runtime.
+
+## Unsupported path formats:
```

Comment:
Should we mention Azure, S3, HTTP, HTTPS, FTP? Or save it for a future edit to remove the focus on GCP?